src/main/java/io/github/edwinmindcraft/apoli/common/power/ModifyPlayerSpawnPower.java

```diff
@@ -1,52 +1,149 @@
+            getSpawn(configuration, entity, serverPlayer.isDeadOrDying());
+        }
+    }
+
+
+    @Override
+    public void onRemoved(ConfiguredPower<ModifyPlayerSpawnConfiguration, ?> configuration, Entity entity) {
+        if (entity instanceof ServerPlayer serverPlayer)
+            if (!serverPlayer.hasDisconnected() && serverPlayer.getRespawnPosition() != null && serverPlayer.isRespawnForced())
+                serverPlayer.setRespawnPosition(Level.OVERWORLD, null, 0F, false, false);
+    }
+
+    // This is not the same tick method because we'd prefer it to only run once.
+    public void tick(Entity entity) {
+        if (entity instanceof ModifyPlayerSpawnCache cache && entity instanceof ServerPlayer spe && cache.getActiveSpawnPower() != null) {
+            IPowerContainer.get(entity).ifPresent(container -> {
+                if ((!container.hasPower(cache.getActiveSpawnPower()) || !container.getPower(cache.getActiveSpawnPower()).isBound() || ApoliAPI.getPowers().containsKey(cache.getActiveSpawnPower()) && !ApoliAPI.getPowers().get(cache.getActiveSpawnPower()).isConfigurationValid() || ApoliAPI.getPowers().containsKey(cache.getActiveSpawnPower()) && !ApoliAPI.getPowers().get(cache.getActiveSpawnPower()).isActive(entity))) {
+                    Optional<ConfiguredPower<ModifyPlayerSpawnConfiguration, ModifyPlayerSpawnPower>> optional = container.getPowers(ApoliPowers.MODIFY_PLAYER_SPAWN.get()).stream().filter(Holder::isBound).map(Holder::value).findFirst();
+                    getSpawn(optional, entity, spe.isDeadOrDying());
+                }
+            });
+        }
+    }
+
+    private static final Set<ServerPlayer> PLAYERS_TO_RESPAWN = Sets.newHashSet();
+
+    public void schedulePlayerToRespawn(ServerPlayer player) {
+        PLAYERS_TO_RESPAWN.add(player);
+    }
 
 	@Override
-	protected void onRespawn(ModifyPlayerSpawnConfiguration configuration, Entity entity) {
-		if (entity instanceof ServerPlayer player && player.getRespawnPosition() == null)
-			this.teleportToModifiedSpawn(configuration, entity);
+	public void onRespawn(ConfiguredPower<ModifyPlayerSpawnConfiguration, ?> configuration, Entity entity) {
+		if (entity instanceof ServerPlayer player && PLAYERS_TO_RESPAWN.contains(player) && player.getRespawnPosition() == null) {
+            this.teleportToModifiedSpawn(configuration, entity);
+            PLAYERS_TO_RESPAWN.remove(player);
+        }
 	}
+
+    private static final ExecutorService FINDER_POOL = Executors.newFixedThreadPool(1);
+
+    public void findSpawn(ConfiguredPower<ModifyPlayerSpawnConfiguration, ?> configuration) {
+        findSpawn(configuration, null);
+    }
+
+    public void findSpawn(ConfiguredPower<ModifyPlayerSpawnConfiguration, ?> configuration, @Nullable ResourceLocation powerId) {
+        SpawnSearchThread searchThread = new SpawnSearchThread(configuration, Optional.ofNullable(powerId));
+        FINDER_POOL.submit(searchThread::start);
```

Comment:
As far as I know, executors work by running the given input in another thread that is allocated by the executor.
This just means the executor will only be able to instantiate a single thread at once, multiple threads may run at a given time.
If you want to use executors, you don't actually need threads.
Here calling ```FINDER_POOL.submit(searchThread::run)``` would do what you expected it to do.

I would also add a config (enabled by default), because while it shouldn't cause any issues, thread safety isn't a guarantee in Minecraft's code. As a rule of thumb, the code could look like
```java
if (ApoliConfig.someValue)
    FINDER_POOL.submit(searchThread::run);
else
    searchThread.run();
```